### DISCUSSION TITLE
How to access a dataframe row?

### DISCUSSION THREAD

**Discussion by Oggy16** [hosseinmoein/DataFrame#246] [Q&A] [answered]
Hi all, 

I'm impressed with the performance improvements of the c++ library vs Python pandas, and I'm in looking to migrate over. 

One question I've had trouble answering is how to access a single row from a dataframe? In pandas you can do this, but this does not work on this library as the dataframe object is not subscriptable.

I've had a look through the docs and while there seem to be a few ways to access columns, I can't find any to access rows. This makes me think it may be something obvious I am missing. 

How do you access an entire row in the dataframe? Best I've found out so far is to access an entire column, and then I can subscript with index.

**Comment by aldebaranzbradaradjan** [comment]
Hi, I still learn how to use it myself (and don't know anything of panda, anyway it's a good lib and it perform very well, thank  hosseinmoein !).
I have had little success with get_row (https://hosseinmoein.github.io/DataFrame/docs/HTML/get_row.html).
I don't have tested from a while, get_row was ok with the index colum, but i could not get any other column. I don't dig in it. I surely don't use it right at the time.
I just access my data by column instead; If i have a moment i will retry get_row to see how it work now, and open an issue if it look like a bug.

**Comment by hosseinmoein** [comment, marked as answer]
There are a couple of ways to access data:
1. You can use the _get_column_ method (https://hosseinmoein.github.io/DataFrame/docs/HTML/get_column.html) which returns a reference to an STL vector. Once you have the reference, you can index through rows or do whatever you could do with an STL vector
2. You can use the _get_row_ method (https://hosseinmoein.github.io/DataFrame/docs/HTML/get_row.html) which returns a heterogeneous vector. See the doc.

I also recommend going through the _hello_world.cc_ file (https://github.com/hosseinmoein/DataFrame/blob/master/examples/hello_world.cc) in some details

**Reply by Oggy16** [reply]
Thank you very much. No idea how I missed that in the documentation. It is a bit more convoluted to do than on Python Pandas, but I put that down to my unfamiliarity with C++. Thank you very much for answering my question!

**Reply by hosseinmoein** [reply]
Yes, unfortunately there is no free lunch. Python is dynamically typed (types are figured out at runtime). That's why Pandas' interface is very elegant and straightforward. C++, on the other hand, is statically typed (all types must be known at compile time). That's why C++ interface is a bit more cluttered. 
Also, that is why Python is slow and not as scalable compared with C++.

**Reply by Oggy16** [reply]
I have to agree, especially on the slowness. When I implemented my simplest data-crunching program for testing, the C++ version with your library used 10x less memory, and was a whopping 168x faster then Python/Pandas for the same task. Considering Pandas is implemented in C I was surprised at how much faster it was.

With improvements like that it is worth for me to learn C++ and deal with the convolutions. I am currently working on re-implementing the rest of the project in C++/DataFrame and I will try to minimise the questions that come up during this period :-)
